File: Data Structures & Algorithms/minimum-stack/submission-0.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_index = -1

    def push(self, val: int) -> None:
        self.stack.append(val)
        if val < self.min_index:
            self.min_index = val

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.stack[self.min_index]

File: Data Structures & Algorithms/daily-temperatures/submission-0.rs
impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut temperatures_copy = temperatures.copy();
        let mut counts = vec![0; temperatures.len()];
        let mds = Vec::new(); // monotonic decreasing stack
        
        for (let i = 0; i < temperatures.len(); i++) {
            // if t > mds.last() {
            //     ...
            // }
            let t = temperatures[i];
            while let Some(&top) = mds.last() {
                if top < t {
                    stack.pop();
                    counts[i] = counts[i] + 1;
                } else {
                    break;
                }
            }
            stack.push(t);
        }

        counts
    }
}
